synapse/handlers/federation.py

```diff
@@ -637,42 +637,70 @@ def __init__(self, hs):
             room_id
             event_ids
 
+        If we fail to fetch any of the events, a warning will be logged, and the event
+        will be omitted from the result. Likewise, any events which turn out not to
+        be in the given room.
+
         Returns:
             map from event_id to event
         """
         fetched_events = await self.store.get_events(event_ids, allow_rejected=True)
 
         missing_events = set(event_ids) - fetched_events.keys()
 
-        if not missing_events:
-            return fetched_events
+        if missing_events:
+            logger.debug(
+                "Fetching unknown state/auth events %s for room %s",
+                missing_events,
+                room_id,
+            )
 
-        logger.debug(
-            "Fetching unknown state/auth events %s for room %s",
-            missing_events,
-            event_ids,
-        )
+            room_version = await self.store.get_room_version(room_id)
 
-        room_version = await self.store.get_room_version(room_id)
+            # XXX 20 requests at once? really?
+            for batch in batch_iter(missing_events, 20):
+                deferreds = [
+                    run_in_background(
+                        self.federation_client.get_pdu,
+                        destinations=[destination],
+                        event_id=e_id,
+                        room_version=room_version,
+                    )
+                    for e_id in batch
+                ]
 
-        # XXX 20 requests at once? really?
-        for batch in batch_iter(missing_events, 20):
-            deferreds = [
-                run_in_background(
-                    self.federation_client.get_pdu,
-                    destinations=[destination],
-                    event_id=e_id,
-                    room_version=room_version,
+                res = await make_deferred_yieldable(
+                    defer.DeferredList(deferreds, consumeErrors=True)
                 )
-                for e_id in batch
-            ]
 
-            res = await make_deferred_yieldable(
-                defer.DeferredList(deferreds, consumeErrors=True)
+                for success, result in res:
+                    if success and result:
+                        fetched_events[result.event_id] = result
+
+        # check for events which were in the wrong room.
+        #
+        # this can happen if a remote server claims that the state or
+        # auth_events at an event in room A are actually events in room B
+
+        bad_events = list(
```

Comment:
```suggestion
        bad_events = (
```

Reply:
no, I'm deliberately building a list here rather than using a generator that refers to `fetched_events`: otherwise we're going to get errors about modifying `fetched_events` while iterating.

Reply:
Wouldn't it be cleaner to have used a list comprehension then rather than calling list on a generator expression?

Reply:
you'll have to fight @hawkowl over that; I'm kinda ambivalent on the topic.

The reason to prefer this syntax (aiui) is for symmetry with constructing a tuple, where you have to invoke `tuple` explicitly to get a tuple rather than a generator.

In the case of creating a list: I'd imagine (I haven't checked) that the bytecode generated for `list(x for x in foo)` is the same as that for `[x for x in foo]` ?

Reply:
It is not - there is no way to guarantee that `list` refers to the list builtin, so it will _always_ compile to a call to `list`.

Reply:
interesting. thanks for the info.